src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs

```diff
@@ -56,66 +56,180 @@ internal static class ReliableStateExtensions
 		/// Query the reliable collection with the given name from the reliable state manager using the given query parameters.
 		/// </summary>
 		/// <param name="stateManager">Reliable state manager for the replica.</param>
+		/// <param name="context">Stateful Service Context.</param>
 		/// <param name="collection">Name of the reliable collection.</param>
 		/// <param name="query">OData query parameters.</param>
 		/// <param name="cancellationToken">Cancellation token.</param>
 		/// <returns>The json serialized results of the query.</returns>
-		public static async Task<IEnumerable<string>> QueryAsync(this IReliableStateManager stateManager, StatefulServiceContext context, string collection, IEnumerable<KeyValuePair<string, string>> query, CancellationToken cancellationToken)
+		public static async Task<IEnumerable<string>> QueryAsync(this IReliableStateManager stateManager,
+			StatefulServiceContext context, string collection, IEnumerable<KeyValuePair<string, string>> query,
+			CancellationToken cancellationToken)
 		{
 			// Query all service partitions concurrently.
 			var proxies = await GetServiceProxiesAsync<IQueryableService>(context).ConfigureAwait(false);
-			var queries = proxies.Select(p => p.QueryPartitionAsync(collection, query)).Concat(new[] { stateManager.QueryPartitionAsync(collection, query, cancellationToken) });
+			var queries = proxies.Select(p => p.QueryPartitionAsync(collection, query)).Concat(new[]
+				{stateManager.QueryPartitionAsync(collection, query, context.PartitionId, cancellationToken)});
 			var queryResults = await Task.WhenAll(queries).ConfigureAwait(false);
 			var results = queryResults.SelectMany(r => r);
 
 			// Run the aggregation query to get the final results (e.g. for top, orderby, project).
 			if (query.Any())
 			{
 				var reliableState = await stateManager.GetQueryableState(collection).ConfigureAwait(false);
-				var valueType = reliableState.GetValueType();
-				var objects = results.Select(r => JsonConvert.DeserializeObject(r, valueType));
-				var queryResult = ApplyQuery(objects, valueType, query, aggregate: true);
+				var entityType = reliableState.GetEntityType();
+				var objects = results.Select(r => JsonConvert.DeserializeObject(r, entityType));
```

Comment:
The code above queries all the other partitions (and gets back a bunch of JSON results).
We deserialize the JSON into objects, re-apply (some of) the query to sort/project/etc.
Then we serialize to JSON, because ServiceProxy serializes things via DataContract and throws an exception if you give it generic objects.

I think we will move to just HTTP on the stateful service side, which will help a bit going forward.